### PR TITLE
Qute: optimize iteration metadata access in loops

### PR DESCRIPTION
**Pull request by mkouba**
This PR is a variation of https://github.com/quarkusio/quarkus/pull/42909/commits/97fd6705c44cab636ff75ee18e387fc105f7be2a. The main difference is that we use the optimization even if metadata prefix is not set.

I gave up the idea of precomputing the metadata keys (mentioned in https://github.com/quarkusio/quarkus/pull/42909/commits/97fd6705c44cab636ff75ee18e387fc105f7be2a#r1738856828) because that would lead to additional 500+ bytes of memory per each loop in a template.

CC @franz1981

### CODE REVIEW

**Comment by mkouba** [comment]
```
RESULTS SUMMARY          |3.8.6                    |3.14.1                   |999-SNAPSHOT             
=========================|Score    |Error  |Diff   |Score    |Error  |Diff   |Score    |Error  |Diff   
-------------------------|-------------------------|-------------------------|-------------------------
HelloParser              |      685|     18|       |      668|     24|    -2%|      649|     35|    -3%
HelloSimple              |    26467|    753|       |    25347|    893|    -4%|    25007|    541|    -1%
IfComplex                |     6121|    114|       |     5790|    324|    -5%|     5606|    357|    -3%
IfSimple                 |     8226|    367|       |     8103|    439|    -1%|     7689|    519|    -5%
IncludeSimple            |    21442|    696|       |    21391|    650|     0%|    20576|    548|    -4%
JavaBeanValueResolver    |    31266|   1704|       |    31175|   1208|     0%|    31194|   1285|     0%
LetComplex               |     5914|    305|       |     5911|    246|     0%|     5765|    297|    -2%
LetSimple                |    16604|    792|       |    16664|    749|     0%|    16748|    381|    +1%
Loop15                   |    33871|   1965|       |    33492|   2216|    -1%|    38086|   2086|   +14%
Loop50                   |    10381|    626|       |    10312|    432|    -1%|    11625|    428|   +13%
NameResolver             |    18454|    620|       |    18514|    442|     0%|    18124|    642|    -2%
When                     |     9148|    454|       |     8892|    577|    -3%|     8848|    505|     0%
```

Hm, a significant improvement in `Loop15`/`Loop50`. The overall picture does not look that good but it could be caused by an "unstable" test environment :thinking:.

**Comment by franz1981** [comment]
@mkouba why not having the version using the string-based switchtable ? did you check if that improve the cases where we regress? 

In my brain, when I did it, was:
- the best possible is the string switch if hashcodes are not computed on the fly AND there's no garbage generation
- with metadata, if there's garbage produced it means both copies cost, fresh hashcode computation (which is very slow for small data, cause is not vectorized "well") - better have the collisions while picking the range length as I did

In short; the latter is suboptimal, perf-wise, but it is a decent trade-off because it doesn't "disturb" the user code by adding unecessary allocations of young objects

**Comment by mkouba** [comment]
> @mkouba why not having the version using the string-based switchtable too ? did you check if that improve the cases where we regress?
> 
> In my brain, when I did it, was:
> 
>     * the best possible is the string switch if hashcodes are not computed on the fly AND there's no garbage generation
> 
>     * with metadata, if there's garbage produced it means both copies cost, fresh hashcode computation (which is very slow for small data, cause is not vectorized "well") - better have the collisions while picking the range length as I did
> 
> 
> In short; the latter is suboptimal, perf-wise, but it is a decent trade-off because it doesn't "disturb" the user code by adding unecessary allocations of young objects

I think that it's better to use a single approach for all use cases, esp. from maintenance POV.

Also the use case without the metadata prefix is more like a corner case and requires additional configuration - the prefix is used by default. So it does not make sense to keep a specialized optimized branch for this corner case IMO.

**Comment by franz1981** [comment]
@mkouba oki, so need to profile the before/after to find out why it appears to have some degradation with the other tests?

**Comment by mkouba** [comment]
> @mkouba oki, so need to profile the before/after to find out why it appears to have some degradation with the other tests?

The bechmarks where the performance dropped do not even contain a loop so I don't think it's related to this PR.

I will run some more benchmarks once all PRs that originated from https://github.com/quarkusio/quarkus/pull/42909 are in. We'll see where the numbers are and if needed analyze the regressions...

**Comment by mkouba** [comment]
> thanks for this work @mkouba 🙏

Thanks for your ideas and PRs!

**Review by gastaldi — APPROVED** [review]

**Review by franz1981 — APPROVED** [review]
LGTM!!

**Review by geoand — APPROVED** [review]

**Comment by geoand** [comment]
Don't worry about the quickstart failure, it has since been fixed
